crates/bindings/src/service.rs

```diff
@@ -0,0 +1,230 @@
+/// Service is a trait implemented by services which may be called from Go.
+pub trait Service {
+    /// Create a new instance of the Service.
+    fn create() -> Self;
+    /// Invoke with the given op-code & |data| payload.
+    /// It extends |arena| with any returned []byte data, and pushes output messages onto |out|.
+    fn invoke(&mut self, code: u32, data: &[u8], arena: &mut Vec<u8>, out: &mut Vec<Out>);
+}
+
+/// Input frame produced from CGO, which is a single service invocation.
+/// 16 bytes, or 1/4 of a typical cache line.
+#[repr(C)]
+pub struct In1 {
+    data_ptr: *const u8,
+    data_len: u32,
+    code: u32,
+}
+
+/// Four invocations, composed into one struct.
+/// 64 bytes, or one typical cache line.
+#[repr(C)]
+pub struct In4 {
+    in0: In1,
+    in1: In1,
+    in2: In1,
+    in3: In1,
+}
+
+/// Sixteen invocations, composed into one struct.
+/// 256 bytes, or four typical cache lines.
+#[repr(C)]
+pub struct In16 {
```

Comment:
I assume there's something preventing you from using `[In1; 16]` instead of the struct. Curious what prevents that from working. Not a big deal, but there's a few places where it'd be cleaner to loop over an array. Might also be able to just use some `unsafe` voodoo to `transmute` this into an array on the rust side, if there's a reason to keep the array out of the C header.

Reply:
Yea, Go needs to be able to access each of these fields. They're individually set by matching implementations on the CGO side.

This song and dance is largely driven by hard requirements of CGO and its needs re: tracking and understanding where every Go pointer is. This forces the ABI to use explicit fields and basically rules out arrays of any kind. [gory details](https://golang.org/cmd/cgo/#hdr-Passing_pointers)

I'd love to, for example, have the Go side pass an array pointer & size of `In1` structs into a single invoke. But, that violates the rule "Go code may pass a Go pointer to C provided the Go memory to which it points does not contain any Go pointers". AKA you can't nest.